frame/stake-tracker/src/mock.rs

```diff
@@ -0,0 +1,265 @@
+	// Staking is the source of truth for voter bags list, since they are not kept up to date.
+	type ScoreProvider = StakingMock;
+	type BagThresholds = BagThresholds;
+	type Score = VoteWeight;
+}
+
+pub struct StakingMock {}
+
+// We don't really care about this yet in the context of testing stake-tracker logic.
+impl ScoreProvider<AccountId> for StakingMock {
+	type Score = VoteWeight;
+
+	fn score(_id: &AccountId) -> Self::Score {
+		VoteWeight::default()
+	}
+
+	frame_election_provider_support::runtime_benchmarks_or_test_enabled! {
+		fn set_score_of(_: &AccountId, _: Self::Score) {
+			// not use yet.
+		}
+	}
+}
+
+impl StakingInterface for StakingMock {
+	type Balance = Balance;
+	type AccountId = AccountId;
+	type CurrencyToVote = frame_support::traits::SaturatingCurrencyToVote;
+
+	fn minimum_nominator_bond() -> Self::Balance {
+		unimplemented!("Currently not used.")
+	}
+
+	fn minimum_validator_bond() -> Self::Balance {
+		unimplemented!("Currently not used.")
+	}
+
+	fn stash_by_ctrl(_: &Self::AccountId) -> Result<Self::AccountId, DispatchError> {
+		unimplemented!("Currently not used.")
+	}
+
+	fn bonding_duration() -> EraIndex {
+		unimplemented!("Currently not used.")
+	}
+
+	fn current_era() -> EraIndex {
+		unimplemented!("Currently not used.")
+	}
+
+	fn stake(
+		who: &Self::AccountId,
+	) -> Result<Stake<Self::AccountId, Self::Balance>, DispatchError> {
+		if *who >= 30 {
+			return Err(DispatchError::Other("not bonded"))
+		}
+		let stake = <Runtime as pallet_stake_tracker::Config>::Currency::total_balance(who);
+		Ok(Stake {
+			stash: *who,
+			active: stake.saturating_sub(ExistentialDeposit::get()),
+			total: stake,
+		})
+	}
+
+	fn bond(_: &Self::AccountId, _: Self::Balance, _: &Self::AccountId) -> DispatchResult {
+		unimplemented!("Currently not used.")
+	}
+
+	fn nominate(_: &Self::AccountId, _: Vec<Self::AccountId>) -> DispatchResult {
+		unimplemented!("Currently not used.")
+	}
+
+	fn chill(_: &Self::AccountId) -> DispatchResult {
+		unimplemented!("Currently not used.")
+	}
+
+	fn bond_extra(_: &Self::AccountId, _: Self::Balance) -> DispatchResult {
+		unimplemented!("Currently not used.")
+	}
+
+	fn unbond(_: &Self::AccountId, _: Self::Balance) -> DispatchResult {
+		unimplemented!("Currently not used.")
+	}
+
+	fn withdraw_unbonded(_: Self::AccountId, _: u32) -> Result<bool, DispatchError> {
+		unimplemented!("Currently not used.")
+	}
+
+	fn desired_validator_count() -> u32 {
+		unimplemented!("Currently not used.")
+	}
+
+	fn election_ongoing() -> bool {
+		unimplemented!("Currently not used.")
+	}
+
+	fn force_unstake(_: Self::AccountId) -> DispatchResult {
+		unimplemented!("Currently not used.")
+	}
+
+	fn is_exposed_in_era(_: &Self::AccountId, _: &EraIndex) -> bool {
+		unimplemented!("Currently not used.")
+	}
+
+	fn is_validator(who: &Self::AccountId) -> bool {
+		*who >= 10 && *who <= 14
+	}
+
+	fn nominations(who: &Self::AccountId) -> Option<Vec<Self::AccountId>> {
+		if *who >= 20 && *who <= 24 {
```

Comment:
If you register your desired validator and nominators using the builder pattern per test, it will be more easier to verify the tests.

Reply:
I've basically made it analogous to how `staking` pallet balances and such were set up, but verifying the tests is indeed a pain, unless you know the pattern to which the setup adheres.

Reply:
You can start by storing a list of nominators directly in a `parameter_types! { pub static Nominators: Vec<_> = .. }`.